### DISCUSSION TITLE
PHPMailer: How to Get Data from HTML Contact Form Sent to Email

### DISCUSSION THREAD

**Discussion by romeyb76** [community/community#108175] [Programming Help]
### Select Topic Area

Question

### Body

I created a mail.php form so when people enter data on my HTML site, then click Submit, the data will get forwarded to my email. I'm using PHPMailer and web server is Hostgator. When I use a test email account and enter text for subject and body, it works, but when I try using the code to pull data from my HTML form, it doesn't work. I get the invalid address to error. Just want to see what I'm doing wrong or if I'm missing something here? Server name, email address and passwords are examples for the purpose of this question.

Here's the HTML

`<form id="contactForm" data-sb-form-api-token="API_TOKEN" action="mail.php" method="POST" enctype="multipart/form-data">
                            
                            <div class="form-floating mb-3">
                                <input class="form-control" id="name" type="text" name="name"  placeholder="Enter your name..." data-sb-validations="required" />
                                <label for="name">Full name</label>
                                <div class="invalid-feedback" data-sb-feedback="name:required">A name is required.</div>
                            </div>
                            
                            <div class="form-floating mb-3">
                                <input class="form-control" id="email" type="email" name="email" placeholder="name@example.com" data-sb-validations="required,email" />
                                <label for="email">Email address</label>
                                <div class="invalid-feedback" data-sb-feedback="email:required">An email is required.</div>
                                <div class="invalid-feedback" data-sb-feedback="email:email">Email is not valid.</div>
                            </div>
                            
                            <div class="form-floating mb-3">
                                <input class="form-control" id="subject" type="text" name="subject" placeholder="Subject" data-sb-validations="required" />
                                <label for="subject">Subject</label>
                                <div class="invalid-feedback" data-sb-feedback="subject:required">A subject is required.</div>
                            </div>
                            
                            <div class="form-floating mb-3">
                                <textarea class="form-control" id="message" type="text" name="message" placeholder="Enter your message here..." style="height: 10rem" data-sb-validations="required"></textarea>
                                <label for="message">Message</label>
                                <div class="invalid-feedback" data-sb-feedback="message:required">A message is required.</div>
                            </div>`

And here's the PHP:

`<?php
//Import PHPMailer classes into the global namespace
//These must be at the top of your script, not inside a function

use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\Exception;

//required files
require 'PHPMailer/src/Exception.php';
require 'PHPMailer/src/PHPMailer.php';
require 'PHPMailer/src/SMTP.php';

//Create an instance; passing `true` enables exceptions

try {
     
    $mail = new PHPMailer(true);
    
    //Server settings
    $mail->SMTPDebug = SMTP::DEBUG_SERVER;                      //Enable verbose debug output
    $mail->isSMTP();                                            //Send using SMTP
    $mail->Host       = 'mail.myemail.com';                     //Set the SMTP server to send through
    $mail->SMTPAuth   = true;                                   //Enable SMTP authentication
    $mail->Username   = 'me@myemail.com';                     //SMTP username
    $mail->Password   = 'SECRET';                               //SMTP password
    $mail->SMTPSecure = 'ssl';            //Enable implicit TLS encryption
    $mail->Port       = 465;                                    //TCP port to connect to; use 587 if you have set `SMTPSecure = PHPMailer::ENCRYPTION_STARTTLS`

    //Recipients
    $mail->setFrom($_POST['email'], $_POST['name']);
    $mail->addAddress('me@myemail.com');     //Add a recipient

    //Content
    $mail->isHTML(true);                                  //Set email format to HTML
    $mail->Subject = $_POST['subject'];
    $mail->Body    = $_POST['message'];
    //  $mail->AltBody = 'This is the body in plain text for non-HTML mail clients';

    $mail->send();
    echo 'Message has been sent';
 
} catch (Exception $e) {
    echo "Message could not be sent. Mailer Error: {$mail->ErrorInfo}";
}
`

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @romeyb76 ! We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Discussions category is solely related to conversations around the GitHub product Discussions. 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Reply by romeyb76** [reply]
Thank you. I didn't know about the categories.

**Comment by lennis-dev** [comment]
Hey @romeyb76,
I think the problem is that in ```$mail->setFrom($_POST['email'], $_POST['name']);``` you are writing from the email address of the sender, but as email you enter your email and as name the one that will be specified in the form you can set as ReplyTo.

I think you need to write the following code instead of ```$mail->setFrom($_POST['email'], $_POST['name']);```
```php
$mail->addReplyTo($_POST['email'], $_POST['name']);
$mail->setFrom("me@myemail.com", $_POST['name']);
```

I hope this solves your problem, if not, please feel free to ask me.

Lennis

**Reply by romeyb76** [reply]
Hi Lennis,

Thank you for your response. I tried that but I keep getting the undefined array key "name" and undefined array key "email" errors. Do you know how I can resolve those?

**Reply by lennis-dev** [reply]
I have tested the code (I have also added a submit button and closing form tag) I have entered my access data and it has worked.

can you possibly provide the whole code (without secrets of course)

**Reply by ucodefusion** [reply]
look like main issue is post data. debug the $_POST

**Comment by ItsJoshBrown** [comment]
I suggest looking into some data sanitization. vs just accepting the user input directory into your application if this is something that is going to be in a Prod environment. 

Here are some resources to help. 
```
https://www.php.net/manual/en/filter.filters.sanitize.php
https://cheatsheetseries.owasp.org/cheatsheets/PHP_Configuration_Cheat_Sheet.html
https://owasp.org
```


You should be able to add something along the lines of this. however, I would just use this as a base and not deploy something like this to your project. 

```PHP
<?PHP

use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\Exception;

require '../vendor/autoload.php';


if ($_SERVER["REQUEST_METHOD"] == "POST") {
    $mail = new PHPMailer(true);


    try {


      //Server settings
        $mail->SMTPDebug = SMTP::DEBUG_SERVER;      //Enable verbose debug output
        $mail->isSMTP();                                                    //Send using SMTP
        $mail->Host = 'SMTP.myemail.com';                       //Set the SMTP server to send through
        $mail->SMTPAuth = true;                                     //Enable SMTP authentication
        $mail->Username = 'me@myemail.com';            //SMTP username
        $mail->Password = 'SECRET';                              //SMTP password
        $mail->SMTPSecure = 'ssl';                                //Enable implicit TLS encryption
        $mail->Port = 465;                                            //TCP port to connect to; use 587 if you have set `SMTPSecure = PHPMailer::ENCRYPTION_STARTTLS`

//Recipients
        $mail->setFrom($_POST['email'], $_POST['name']);
        $mail->addAddress('me@myemail.com');     //Add a recipient

//Content
        $mail->isHTML(true);                                  //Set email format to HTML
        $mail->Subject = $_POST['subject'];
        $mail->Body = $_POST['message'];
//  $mail->AltBody = 'This is the body in plain text for non-HTML mail clients';

        $mail->send();
        echo 'Message has been sent';

    } catch (Exception $e) {
        echo "Message could not be sent. Mailer Error: {$mail->ErrorInfo}";
    }
}
?>
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Contact</title>
  </head>
  <body>
    <form id="contactForm" data-sb-form-api-token="API_TOKEN" action="mail.php" method="POST" enctype="multipart/form-data">
      <div class="form-floating mb-3">
        <input class="form-control" id="name" type="text" name="name" placeholder="Enter your name..." data-sb-validations="required" />
        <label for="name">Full name</label>
        <div class="invalid-feedback" data-sb-feedback="name:required">A name is required.</div>
      </div>
      <div class="form-floating mb-3">
        <input class="form-control" id="email" type="email" name="email" placeholder="name@example.com" data-sb-validations="required,email" />
        <label for="email">Email address</label>
        <div class="invalid-feedback" data-sb-feedback="email:required">An email is required.</div>
        <div class="invalid-feedback" data-sb-feedback="email:email">Email is not valid.</div>
      </div>
      <div class="form-floating mb-3">
        <input class="form-control" id="subject" type="text" name="subject" placeholder="Subject" data-sb-validations="required" />
        <label for="subject">Subject</label>
        <div class="invalid-feedback" data-sb-feedback="subject:required">A subject is required.</div>
      </div>
      <div class="form-floating mb-3">
        <textarea class="form-control" id="message" type="text" name="message" placeholder="Enter your message here..." style="height: 10rem" data-sb-validations="required"></textarea>
        <label for="message">Message</label>
        <div class="invalid-feedback" data-sb-feedback="message:required">A message is required.</div>
      </div>
      <button type="submit" class="btn btn-primary">Submit</button>
    </form>
  </body>
</html>```

**Comment by ethanappgenius** [comment]
The issue you're encountering with PHPMailer may be due to how you're setting the sender's email address. In your PHP script, you're using `$_POST['email']` directly without proper validation, which can lead to security vulnerabilities like email header injection. Instead, ensure that you sanitize and validate user inputs before using them in your email headers. Additionally, make sure that the email address provided by the user through the form is a valid email address. You can use PHP's built-in `filter_var()` function with the `FILTER_VALIDATE_EMAIL` filter to validate the email address. By validating and sanitizing user inputs properly, you can prevent errors and enhance the security of your application.

**Comment by mecodeatlas** [comment]
Hey @romeyb76 👋 

We're going through discussions that haven't been marked as answered yet. Did the replies from other members help you resolve the issue you raised? If so, can you please mark their response as the answer? When you mark a question as an answer, GitHub will highlight the comment and replies to the comment to help visitors quickly find the answer.
![image](https://github.com/user-attachments/assets/b89e7161-c2d1-47e2-a058-242df633e6b2)

**Comment by mecodeatlas** [comment]
Hi @romeyb76! We hope you found what you were looking for. ✨

We're closing this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
